src/webots/gui/WbStreamingServer.cpp

```diff
@@ -190,54 +190,44 @@ void WbStreamingServer::onNewTcpConnection() {
-  if (tokens[0] == "GET") {
-    const QString &requestedUrl(tokens[1].replace(QRegularExpression("^/"), ""));
-    if (!requestedUrl.isEmpty()) {  // "/" is reserved for the websocket.
-      bool hasEtag = false;
-      QString etag;
-      for (const auto &i : tokens) {
-        if (i == "If-None-Match:")
-          hasEtag = true;
-        else if (hasEtag) {
-          etag = i;
-          break;
-        }
-      }
-      sendTcpRequestReply(requestedUrl, etag, socket);
-    }
+  if (tokens[0] == "GET" && tokens[1] != "/") {  // "/" is reserved for the websocket.
+    const int hostIndex = tokens.indexOf("Host:") + 1;
+    const QString host = hostIndex ? tokens[hostIndex] : "";
+    const int etagIndex = tokens.indexOf("If-None-Match:") + 1;
+    const QString etag = etagIndex ? tokens[etagIndex] : "";
+    if (host.isEmpty())
+      WbLog::warning(tr("No host specified in HTTP header."));
+    sendTcpRequestReply(tokens[1].sliced(1), etag, host, socket);
   }
 }
 
-void WbStreamingServer::sendTcpRequestReply(const QString &completeUrl, const QString &etag, QTcpSocket *socket) {
-  QString requestedUrl = completeUrl.left(completeUrl.lastIndexOf('?'));
-  QString filePath = WbProject::current()->pluginsPath() + requestedUrl;
-  // Here handle the streaming_viewer files.
-  static const QStringList streamer_files = {"index.html", "setup_viewer.js", "style.css", "webots_icon.png"};
-  if ((!requestedUrl.startsWith("robot_windows/")) && streamer_files.contains(requestedUrl)) {
-    if (streamer_files.contains(requestedUrl))
-      filePath = WbStandardPaths::resourcesWebPath() + "streaming_viewer/" + requestedUrl;
-    else {
-      WbLog::warning(tr("Unsupported URL %1").arg(requestedUrl));
-      socket->write(WbHttpReply::forge404Reply());
-    }
-  } else if (requestedUrl.contains("generic")) {
-    QString requestFile = requestedUrl.mid(requestedUrl.lastIndexOf("/"));
-    filePath = WbStandardPaths::resourcesRobotWindowsPluginsPath() + "generic" + requestFile;
-  } else if (requestedUrl.endsWith(".js")) {
+void WbStreamingServer::sendTcpRequestReply(const QString &completeUrl, const QString &etag, const QString &host,
+                                            QTcpSocket *socket) {
+  const QString requestedUrl = completeUrl.left(completeUrl.lastIndexOf('?'));
+  QString filePath;
+  static const QStringList streamerFiles = {"index.html", "setup_viewer.js", "style.css", "webots_icon.png"};
+  if (requestedUrl == "streaming")  // shortcut
+    filePath = WbStandardPaths::resourcesWebPath() + "streaming_viewer/index.html";
+  else if (streamerFiles.contains(requestedUrl))
+    filePath = WbStandardPaths::resourcesWebPath() + "streaming_viewer/" + requestedUrl;
+  else if (requestedUrl.startsWith("robot_windows/"))
+    filePath = WbProject::current()->pluginsPath() + requestedUrl;
+  else if (requestedUrl.startsWith("resources/projects/plugins/robot_windows/generic/") || requestedUrl.endsWith(".js") ||
```

Comment:
This seems wrong, at least when I try to open a generic robot window, it does not find it.
But it bugs only in the development environment.

Reply:
I believe I fixed it in a simple way, without breaking any backwards compatibility.
However, I will open a new issue about URLs for robot windows and robot controllers, that should behave like URL for images, sounds and meshes, with either relative or absolute references as discussed.